Introducing ChecksumType

Nimmaster opened the pull request:
To represent well-known checksum algorithms as enum (to not to use magic strings or constants) and encapsulate checksum file logic in one class.

amihaiemil commented:
@g4s8 Can I have this one?

0crat commented:
This pull request [#14](https://github.com/artipie/maven-adapter/issues/14) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/214aa483-6bc4-480e-9a10-f2f8e4b8b92f); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

<!-- https://www.0crat.com/footprint/CT2E6TK9B/c0f94cfc-4024-4ead-996a-bba10d2d13c8, version: 0.51.1, hash: ${buildNumber} -->

g4s8 commented:
@0crat assign @amihaiemil

0crat commented:
> @0crat assign @amihaiemil  ([here](https://github.com/artipie/maven-adapter/issues/14#issuecomment-582390348))

@g4s8 There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 11@8af595d0a0c8, thread: PQ-CT2E6TK9B
com.zerocracy.claims.ClaimIn[214] java.lang.IllegalArgumentException: Parameter "age" not found in "Order was canceled" among: [comment, job, issue, repo, cause, _parent_sig, flow, login]
```

[0.51.1](https://github.com/zerocracy/farm): CID: [681b15c4-cd72-451d-97f8-24264b2dc60e](https://www.0crat.com/footprint/CT2E6TK9B/681b15c4-cd72-451d-97f8-24264b2dc60e), Type: "Order was canceled"

<!-- https://www.0crat.com/footprint/CT2E6TK9B/e39fca6a-3d04-4f98-b0f9-eeeddf47f7cb, version: 0.51.1, hash: ${buildNumber} -->

amihaiemil commented:
@Nimmaster Also, as usual, PLEASE specify a description for the PR. Link to the Issue and explain what does this PR introduce...

amihaiemil commented:
@Nimmaster Also think about tests. If you don't have the time to write unit tests, plese leave a TODO (puzzle).

amihaiemil commented:
@Nimmaster About the ``digest(...)`` method(s). I think maybe they should be part of class ``Algorithm``? 

In which case you don't need the Enum at all. You have objects of type Algorithm, encapsulating the type of the algorithm and providing the proper ``Digest`` implementation, via the ``digest(...)`` method(s).

0crat commented:
> @0crat assign @amihaiemil  ([here](https://github.com/artipie/maven-adapter/issues/14#issuecomment-582390348))

@g4s8 This pull request [#14](https://github.com/artipie/maven-adapter/issues/14) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/fccccb62-1721-4424-a580-6b86663c8b04); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

<!-- https://www.0crat.com/footprint/CT2E6TK9B/fdb06dbe-ff46-4664-b364-a10b6a1cf33c, version: 0.51.1, hash: ${buildNumber} -->

0crat commented:
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

<!-- https://www.0crat.com/footprint/CT2E6TK9B/885d6a91-750e-4010-96c2-184ab2371a22, version: 0.51.1, hash: ${buildNumber} -->

Nimmaster commented:
@amihaiemil 
I refactored it. Basically I don't really need generic classes for hashing or its result, so I added `ChecksumAttribute` class which is more domain oriented. It encapsulates checksum calculation and createion of an attribute file (a file with checksum algorithm as extension and containing hex-encoded checksum).

I need to support just two algorithms, so it's natural to represent it as Java Enum.

Nimmaster commented:
@amihaiemil I made it green again.

fanifieiev commented:
@Nimmaster @amihaiemil 
Here are my honest concerns:
1. The methods' names in `ChecksumAttribute`  class are bad named according to `Elegant Objects principles`. 
Take a look at [Builders and Manipulators](https://www.yegor256.com/2018/08/22/builders-and-manipulators.html) 
2. There are two many methods in `ChecksumAttribute`. From the user perspective I do not really understand what each method does, I think there should only be one method(ideally) and the class itself should implement interface.
3. All the code done in this PR is available in [Cactoos ](https://github.com/yegor256/cactoos) project. 
@g4s8  Why not introducing Cactoos dependency into this project???
 Take a look at: 
[Sha1DigestOf](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/io/Sha1DigestOf.java)
[Sha256DigestOf](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/io/Sha256DigestOf.javaurl)
[Md5DigestOf](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/io/Md5DigestOf.java)
[HexOf](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/bytes/HexOf.java)
and bunch of other bytes and IO featured classes in [bytes
](https://github.com/yegor256/cactoos/tree/master/src/main/java/org/cactoos/bytes) and [IO](https://github.com/yegor256/cactoos/tree/master/src/main/java/org/cactoos/io)

amihaiemil commented:
@Nimmaster also please don't forget about the PR description

g4s8 commented:
@Nimmaster @amihaiemil yes, cactoos already has classes to read hex of MD5 and SHA1 from file, e.g. using new `HexOf(new Md5DigestOf(new InputOf(file))).asString()`

yegor256 commented on src/main/java/com/artipie/maven/ChecksumAttribute.java:
```diff
@@ -0,0 +1,123 @@
+                StandardOpenOption.CREATE
+            );
+            return hex;
+        } catch (final IOException ex) {
```
@Nimmaster just let it fly to the highest level, the code will be shorter here

Nimmaster replied:
I still prefer to localize checked exceptions instead of forcing class users to catch it.

yegor256 replied:
@Nimmaster I realize this, but in artipie projects we do it differently :)

yegor256 commented on src/main/java/com/artipie/maven/ChecksumAttribute.java:
```diff
@@ -0,0 +1,123 @@
+            final var digest = MessageDigest.getInstance(type.algorithm());
+            read.transferTo(new DigestOutputStream(new NullOutputStream(), digest));
+            return Hex.encodeHexString(digest.digest());
+        } catch (final IOException ex) {
```
@Nimmaster let it fly up

yegor256 commented on src/main/java/com/artipie/maven/ChecksumAttribute.java:
```diff
@@ -0,0 +1,123 @@
+     * @param type Checksum algorithm
+     * @return HEX-encoded checksum
+     */
+    String readHex(final ChecksumType type) {
```
@Nimmaster package-level visibility is a bad practice in general, make it either `private` or `public`

amihaiemil commented:
@Nimmaster Please also let the exceptions fly up. I am also an addept of "fail-fast" principles. We should not swallow any exceptions :)

amihaiemil commented:
@g4s8 So if Cactoos already offers this functionality, does it make sense to merge this PR, or...?

amihaiemil commented:
@Nimmaster And I ask you for the third time, please provide a description for the PR, and link the corresponding Issue.

Nimmaster commented:
@amihaiemil @g4s8 if the file is being deleted then it's already no longer needed. We should not fail in this case

amihaiemil commented:
@Nimmaster But what do you do if the deletion fails? Move on and do nothing? We should actually have a Retry decorator to retry the operation a few times if it fails.

g4s8 commented:
@Nimmaster 
> We should not fail in this case

This decision should be made by client code, not by object itself. Client may ignore this error or perform some actions.

Nimmaster commented:
@amihaiemil @g4s8 That comment belonged to [another pull request](https://github.com/artipie/maven-adapter/pull/16) sorry. I fixed it there BTW.

Runtime exceptions are automatically propagated if they were not caught. Checked exceptions give you nothing, just make you write more boilerplate.

amihaiemil commented:
@Nimmaster  I don't have anything to do further here, I think. You should come to an agreement with @g4s8 , he is the architect.

amihaiemil commented:
@0crat status

0crat commented:
> @0crat status ([here](https://github.com/artipie/maven-adapter/issues/14#issuecomment-583268237))

@amihaiemil This is what I know about this job in [CT2E6TK9B](https://www.0crat.com/p/CT2E6TK9B), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#14](https://github.com/artipie/maven-adapter/issues/14) is [in scope](http://www.zerocracy.com/policy.html#14) for 2days
  * The role is `REV`
  * The job is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) for [2days](http://www.zerocracy.com/policy.html#8)
  * There is a monetary reward attached
  * @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) will get 25% of their hourly rate for this job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @nimmaster[/z](https://www.0crat.com/u/nimmaster): This user reported the ticket
    * @fanifieiev[/z](https://www.0crat.com/u/fanifieiev): User was resigned from the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fmaven-adapter%2314%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

<!-- https://www.0crat.com/footprint/CT2E6TK9B/d2c61058-4158-49f7-b243-63e58ad5a55a, version: 0.51.1, hash: ${buildNumber} -->

amihaiemil commented:
@g4s8 can you force the bot to make the payment here? :D

g4s8 commented:
@0crat out

amihaiemil commented:
@g4s8 Still no payment here. Could there be a problem, because I changed the PR's title? I hope now :(

0crat commented:
Order was finished: +15 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

<!-- https://www.0crat.com/footprint/CT2E6TK9B/5c549644-faf7-4fa4-8a69-8f21cb42ea81, version: 0.51.1, hash: ${buildNumber} -->

0crat commented:
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

<!-- https://www.0crat.com/footprint/CT2E6TK9B/ef4b3c77-18bc-4629-9ea7-b5a075829794, version: 0.51.1, hash: ${buildNumber} -->
